{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import time\n",
    "\n",
    "consensus = []\n",
    "range_variation = []\n",
    "nucleotide_count =[]\n",
    "plocation = []\n",
    "i_plocation = []\n",
    "l_plocation = []\n",
    "\n",
    "# Range for creating plotting colors\n",
    "range1 = 50\n",
    "range2 = 500\n",
    "range3 = 2500\n",
    "\n",
    "# Mutation threshold\n",
    "m_t = 50\n",
    "\n",
    "sizeb = 0\n",
    "\n",
    "fip = 0\n",
    "rfip = 0\n",
    "a_end = 0\n",
    "\n",
    "def ComputeMutation (p, nc_i ,i):\n",
    "    if n > m_t:\n",
    "        \n",
    "        mutated_nt.seq[p] += \" \" + str(nc_i-1) + \" \"\n",
    "        output_s.seq[p] = output_s.seq[p] + \"\\n\" + str(nc_i) + \"\\t\" + str(n) + \"\\t\" + \"\\t\".join(nucleotide_count[i].split('\\t')[1:6])\n",
    "\n",
    "\n",
    "        \n",
    "class sequences:\n",
    "    def __init__(self,p):\n",
    "        self.seq = {\n",
    "            \"F3\":p,\n",
    "            \"B3\":p,\n",
    "            \"LF\":p,\n",
    "            \"LB\":p,\n",
    "            \"FIP\":p,\n",
    "            \"rFIP\":p,\n",
    "            \"BIP\":p,\n",
    "            \"rBIP\":p\n",
    "        }\n",
    "\n",
    "p_start = sequences(0)\n",
    "primers = sequences(\"\")\n",
    "primers_fragments = sequences(\"\")\n",
    "output_s = sequences(\"\")\n",
    "mutated_nt = sequences(\"\")\n",
    "\n",
    "\n",
    "print(\"________________ALIGNING PRIMERS_________________\")\n",
    "with open(\"consensusSeq.txt\", 'r') as cons:\n",
    "    for record in SeqIO.parse(cons, \"fasta\"):\n",
    "        consensus = str(record.seq).upper()\n",
    "    with open(\"primers.txt\", 'r') as primersIO:\n",
    "            for primersIO in SeqIO.parse(primersIO, \"fasta\"):\n",
    "                if \"F3\" in primersIO.id:\n",
    "                    \n",
    "                    primers.seq[\"F3\"] = primersIO.seq\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq,2,-4, -10,-1)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    p_start.seq[\"F3\"] = alignments[0][3]\n",
    "                    primers_fragments.seq[\"F3\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "\n",
    "                elif \"B3\" in primersIO.id:\n",
    "                    \n",
    "                    primers.seq[\"B3\"] = primersIO.seq\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq.reverse_complement(),2,-4, -10,-1)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    p_start.seq[\"B3\"] = alignments[0][3]\n",
    "                    primers_fragments.seq[\"B3\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "                    \n",
    "                    a_end = alignments[0][4]\n",
    "                    \n",
    "                if \"LF\" in primersIO.id:\n",
    "                    \n",
    "                    primers.seq[\"LF\"] = primersIO.seq\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq.reverse_complement(),2,-4, -10,-1)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    p_start.seq[\"LF\"] = alignments[0][3]\n",
    "                    primers_fragments.seq[\"LF\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "\n",
    "                elif \"LB\" in primersIO.id:\n",
    "                    \n",
    "                    primers.seq[\"LB\"] = primersIO.seq\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq,2,-4, -10,-1)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    p_start.seq[\"LB\"] = alignments[0][3]\n",
    "                    primers_fragments.seq[\"LB\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "\n",
    "                elif \"FIP\" in primersIO.id:\n",
    "                    \n",
    "                    primers.seq[\"FIP\"] = primersIO.seq\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq,2,-4, -10,-1)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    p_start.seq[\"FIP\"] = alignments[0][3]\n",
    "                    buff = format_alignment(*alignments[0]).split(\" \")\n",
    "                    primers_fragments.seq[\"FIP\"] = buff[1].split(\"\\n\")[0]\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq.reverse_complement(),2,-4, -10,-1)\n",
    "                    print(\"----------REVERSE COMPLEMENT OF \" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    buff = format_alignment(*alignments[0]).split(\" \")\n",
    "                    p_start.seq[\"rFIP\"] = alignments[0][3]\n",
    "                    primers_fragments.seq[\"rFIP\"] = buff[1].split(\"\\n\")[0]\n",
    "                    \n",
    "\n",
    "                elif \"BIP\" in primersIO.id:\n",
    "                    \n",
    "                    primers.seq[\"BIP\"] = primersIO.seq\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq,2,-4, -10,-1)\n",
    "                    print(\"----------REVERSE COMPLEMENT OF \" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    p_start.seq[\"rBIP\"] = alignments[0][3]\n",
    "                    primers_fragments.seq[\"rBIP\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "                    sizeb=len(primersIO.seq)\n",
    "                    \n",
    "                    alignments = pairwise2.align.localms(consensus, primersIO.seq.reverse_complement(),2,-4, -10,-1)\n",
    "                    print(\"----------\" + primersIO.id + \"----------\")\n",
    "                    print(format_alignment(*alignments[0]))\n",
    "                    p_start.seq[\"BIP\"] = alignments[0][3]\n",
    "                    primers_fragments.seq[\"BIP\"] = format_alignment(*alignments[0]).split(\" \")[1].split(\"\\n\")[0]\n",
    "            \n",
    "print(\"PRIMERS RANGE:\" + str(p_start.seq[\"F3\"]) + \" - \" + str(a_end))\n",
    "\n",
    "print(\"________________COMPUTING VARIATION MATRIX_________________\")\n",
    "\n",
    "if not os.path.isdir(\"LampPrimers\"):\n",
    "    os.mkdir(\"LampPrimers\")\n",
    "f_name = int(round(time.time() * 1000))\n",
    "os.mkdir(\"LampPrimers/%s\" % (f_name))\n",
    "\n",
    "with open(\"LampPrimers/\" + str(f_name) + \"/output.txt\", 'w') as output:\n",
    "    with open(\"Variation.txt\", 'r') as variation:\n",
    "        with open(\"nc_count.txt\", 'r') as ncount:\n",
    "\n",
    "            nc_variation = variation.readline()\n",
    "            ncount.readline()\n",
    "#             The +1 in the following line is because pairwise2 alignment bases number start at 0 and the nc_count & variation.txt starts at 1\n",
    "            while not str(p_start.seq[\"F3\"]+1) in nc_variation:\n",
    "                nc_variation = variation.readline()\n",
    "                ncount.readline()\n",
    "\n",
    "            range_variation.append(float(nc_variation.split(\"\\t\")[1].split(\"\\n\")[0]))\n",
    "            nucleotide_count.append(ncount.readline())\n",
    "\n",
    "            while not str(a_end) in nc_variation:\n",
    "                \n",
    "                \n",
    "                nc_variation = variation.readline()\n",
    "                range_variation.append(float(nc_variation.split(\"\\t\")[1].split(\"\\n\")[0]))\n",
    "                \n",
    "                nucleotide_count.append(ncount.readline())\n",
    "                \n",
    "#                 print(nucleotide_count[len(nucleotide_count)-1])\n",
    "#                 print(range_variation[len(nucleotide_count)-1])\n",
    "\n",
    "            buff = []\n",
    "            i = 0\n",
    "\n",
    "            for n,i in zip(range_variation, range(0,len(range_variation))):\n",
    "                #calculate variation\n",
    "                if n < range1:\n",
    "                    buff.append(0)\n",
    "                elif n < range2:\n",
    "                    buff.append(0.33)\n",
    "                elif n < range3:\n",
    "                    buff.append(0.66)\n",
    "                else:\n",
    "                    buff.append(1)\n",
    "\n",
    "                #Do this region have primers?\n",
    "            #             F3 & B3\n",
    "                if i < len(str(primers_fragments.seq[\"F3\"])): #F3\n",
    "                    plocation.append(0.1)\n",
    "                    nucleotide_index = i+1\n",
    "                    ComputeMutation(\"F3\", nucleotide_index, i)\n",
    "\n",
    "\n",
    "                elif i > (p_start.seq[\"B3\"]-p_start.seq[\"F3\"]) and i < ((p_start.seq[\"B3\"]-p_start.seq[\"F3\"]) + len(str(primers_fragments.seq[\"B3\"]))):\n",
    "                    plocation.append(0.9)\n",
    "                    nucleotide_index = len(primers_fragments.seq[\"B3\"])-(i-(p_start.seq[\"B3\"]-p_start.seq[\"F3\"]))\n",
    "                    ComputeMutation(\"B3\", nucleotide_index, i)\n",
    "\n",
    "                else:\n",
    "                    plocation.append(0)\n",
    "\n",
    "\n",
    "            #                 FIP AND BIP\n",
    "                if i > (p_start.seq[\"FIP\"]-p_start.seq[\"F3\"]) and i < ((p_start.seq[\"FIP\"]-p_start.seq[\"F3\"]) + len(str(primers_fragments.seq[\"FIP\"]))): \n",
    "                    i_plocation.append(0.3)\n",
    "                    nucleotide_index = len(primers.seq[\"FIP\"]) - (len(primers_fragments.seq[\"FIP\"])- ((i+1) - (p_start.seq[\"FIP\"] - p_start.seq[\"F3\"])))\n",
    "                    ComputeMutation(\"FIP\", nucleotide_index, i)\n",
    "\n",
    "                elif i > (p_start.seq[\"rFIP\"]-p_start.seq[\"F3\"]) and i < ((p_start.seq[\"rFIP\"]-p_start.seq[\"F3\"]) + len(str(primers_fragments.seq[\"rFIP\"]))): \n",
    "                    i_plocation.append(0.3)\n",
    "                    nucleotide_index = (p_start.seq[\"rFIP\"]-p_start.seq[\"F3\"]) + len(primers_fragments.seq[\"rFIP\"]) - i\n",
    "                    if nucleotide_index > len(primers_fragments.seq[\"rBIP\"]):\n",
    "                        ComputeMutation(\"rFIP\", nucleotide_index, i)\n",
    "\n",
    "                elif i > (p_start.seq[\"rBIP\"]-p_start.seq[\"F3\"]) and i < ((p_start.seq[\"rBIP\"]-p_start.seq[\"F3\"]) + len(str(primers_fragments.seq[\"rBIP\"]))): \n",
    "                    i_plocation.append(0.6)\n",
    "                    nucleotide_index = (i+1)-(p_start.seq[\"rBIP\"]-p_start.seq[\"F3\"])\n",
    "                    ComputeMutation(\"rBIP\", nucleotide_index, i)\n",
    "\n",
    "                elif i > (p_start.seq[\"BIP\"]-p_start.seq[\"F3\"]) and i < ((p_start.seq[\"BIP\"]-p_start.seq[\"F3\"]) + len(str(primers_fragments.seq[\"BIP\"]))): \n",
    "                    i_plocation.append(0.6)\n",
    "                    nucleotide_index = len(primers.seq[\"BIP\"]) - (i - (p_start.seq[\"BIP\"]-p_start.seq[\"F3\"]))\n",
    "                    if nucleotide_index > len(primers_fragments.seq[\"rBIP\"]):\n",
    "                        ComputeMutation(\"BIP\", nucleotide_index, i)\n",
    "\n",
    "                else:\n",
    "                    i_plocation.append(0)\n",
    "\n",
    "\n",
    "\n",
    "            #                 LOOP PRIMERS   \n",
    "                if i > (p_start.seq[\"LF\"]-p_start.seq[\"F3\"]) and i < ((p_start.seq[\"LF\"]-p_start.seq[\"F3\"]) + len(str(primers_fragments.seq[\"LF\"]))):\n",
    "                    l_plocation.append(0.5)\n",
    "                    nucleotide_index = len(primers_fragments.seq[\"LF\"])-(i-(p_start.seq[\"LF\"]-p_start.seq[\"F3\"]))\n",
    "                    ComputeMutation(\"LF\", nucleotide_index, i)\n",
    "\n",
    "                elif i > (p_start.seq[\"LB\"]-p_start.seq[\"F3\"]) and i < ((p_start.seq[\"LB\"]-p_start.seq[\"F3\"]) + len(str(primers_fragments.seq[\"LB\"]))):\n",
    "                    l_plocation.append(0.5)\n",
    "                    nucleotide_index = 1+i-(p_start.seq[\"LB\"]-p_start.seq[\"F3\"])\n",
    "                    ComputeMutation(\"LB\", nucleotide_index, i)\n",
    "                else:\n",
    "                    l_plocation.append(0)\n",
    "\n",
    "\n",
    "            range_variation = buff\n",
    "\n",
    "            rv = np.array([range_variation])\n",
    "            rv = np.vstack((rv, np.array([plocation])))\n",
    "            rv = np.vstack((rv, np.array([i_plocation])))\n",
    "            rv = np.vstack((rv, np.array([l_plocation])))\n",
    "\n",
    "            fig, ax = plt.subplots(figsize=( (a_end - p_start.seq[\"F3\"])/10, 1))\n",
    "            im = ax.imshow(rv,aspect='auto')\n",
    "\n",
    "\n",
    "            plt.show()\n",
    "            fig.savefig(\"LampPrimers/\" + str(f_name) + \"/heatmap.png\")\n",
    "            \n",
    "            for n in primers.seq:\n",
    "                if primers.seq[n] != \"\":\n",
    "                    output.write(\">\" + n +\"\\n\")\n",
    "                    output.write(str(primers.seq[n]) + \"\\n\")\n",
    "            \n",
    "            output.write(\"\\n------------------------------------------------\")\n",
    "            output.write(\"\\n|            MUTATION ANALYSIS                 |\")\n",
    "            output.write(\"\\n------------------------------------------------\\n\")\n",
    "            \n",
    "#             Calculate FIP and BIP complete sequences\n",
    "            \n",
    "            output_s.seq[\"FIP\"] = output_s.seq[\"rFIP\"] + output_s.seq[\"FIP\"]\n",
    "            output_s.seq[\"BIP\"] = output_s.seq[\"BIP\"] + output_s.seq[\"rBIP\"]\n",
    "            \n",
    "            mutated_nt.seq[\"FIP\"] = mutated_nt.seq[\"rFIP\"] + mutated_nt.seq[\"FIP\"]\n",
    "            mutated_nt.seq[\"BIP\"] = mutated_nt.seq[\"BIP\"] + mutated_nt.seq[\"rBIP\"]\n",
    "        \n",
    "            for n in output_s.seq:\n",
    "                        \n",
    "                if output_s.seq[n] != \"\" and not \"r\" in n:\n",
    "                    \n",
    "                    output.write(\"\\n\" + str(n) + \"\\n\")\n",
    "                    output.write(\"------------------------------------------\\n\")\n",
    "                    for nc,ni in zip(primers.seq[n],range(0,len(primers.seq[n]))):\n",
    "                        if \" \" + str(ni) + \" \" in mutated_nt.seq[n]:\n",
    "                            output.write(\" *\" + nc + \"* \")\n",
    "                        else:\n",
    "                            output.write(nc)\n",
    "                    output.write(\"\\nMutated nucleotides count in the TARGET STRAND:\")\n",
    "                    output.write(\"\\nnc\\tvar.\\tA\\tT\\tG\\tC\\t-\\n\")\n",
    "                    output.write(output_s.seq[n])\n",
    "                \n",
    "            output.close()\n",
    "            print(\"Row legend:\\n\\t- 1st: Mutation rate.\\n\\t- 2nd: B3 and F3 primers alignment.\\n\\t- 3rd: FIP and BIP primers alignment.\\n\\t- 4th: Loop primers alignment.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
